File: 스택_큐/지현/다리를_지나는_트럭.py

def solution(bridge_length, weight, truck_weights):
    bridge=[]
    for i in range(bridge_length):
        bridge.append(0)
    # bridge=[0 for i in range(bridge_length)] #또다른 방법
    count=0
    sumWeight=0
    while truck_weights!=[]:
        count+=1
        sumWeight-=bridge[-1]
        del bridge[-1]
        
        bridge.insert(0,0)
        #원래는 sum(bridge)로 다리위 무게총합을 구했는데 그렇게하니 시간초과 발생함 
        if sumWeight+truck_weights[0]<=weight:
            bridge[0]=truck_weights[0]
            sumWeight+=truck_weights[0]
            del truck_weights[0]

    return count+bridge_length

print(solution(2,10,[7, 4, 5, 6]))

File: 스택_큐/[희진]스택_큐/다리를_지나는_트럭.js
function solution(bridge_length, weight, truck_weights) {
  // 다리를 나타내는 큐를 초기화하고, 초기 상태는 모두 0으로 채움
  const bridgeQueue = Array(bridge_length).fill(0);

  // 현재 다리 위의 무게를 나타내는 변수 초기화
  let currentWeight = 0;

  // 경과 시간을 나타내는 변수 초기화
  let time = 0;

  // 다리 위에 트럭이 있거나, 대기 중인 트럭이 남아있을 때까지 반복
  while (bridgeQueue.length > 0 || truck_weights.length > 0) {
    // 시간이 1초씩 증가
    time++;

    // 다리를 나가는 트럭의 무게를 현재 다리 위의 무게에서 뺌
    currentWeight -= bridgeQueue.shift();

    // 대기 중인 트럭이 다리에 올라갈 수 있을 때
    if (
      truck_weights.length > 0 &&
      currentWeight + truck_weights[0] <= weight
    ) {
      // 대기 중인 트럭 중 가장 앞의 트럭을 꺼내 다리에 넣고, 다리 위 무게를 갱신
      const truck = truck_weights.shift();
      bridgeQueue.push(truck);
      currentWeight += truck;
    } else {
      // 대기 중인 트럭이 다리에 올라갈 수 없으면 0을 큐에 추가
      bridgeQueue.push(0);
    }
  }

  // 최종적으로 걸린 시간을 반환
  return time;
}
